The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11460433349/job/31887150232) [(plain)](https://github.com/rust-lang/rust/commit/9e383ddd46ffe69b8dd3aeb830642b7a5909baa7/checks/31887150232/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.755 Building wheels for collected packages: reuse
#13 2.756   Building wheel for reuse (pyproject.toml): started
#13 3.004   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.005   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.005   Stored in directory: /tmp/pip-ephem-wheel-cache-scouswr7/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.008 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.408 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.408 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.941 Collecting virtualenv
#13 3.941 Collecting virtualenv
#13 3.981   Downloading virtualenv-20.27.0-py3-none-any.whl (3.1 MB)
#13 4.049      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 47.2 MB/s eta 0:00:00
#13 4.087 Collecting distlib<1,>=0.3.7
#13 4.093   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.102      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 71.5 MB/s eta 0:00:00
#13 4.136 Collecting filelock<4,>=3.12.2
#13 4.140   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.174 Collecting platformdirs<5,>=3.9.1
#13 4.178   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.258 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.437 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.0
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      180160 kB
DirectMap2M:     6111232 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARNING: `rust.download-rustc` is enabled. The `rust.channel` option will be overridden by the CI rustc's channel.
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
error: missing type for `const` item
##[error]  --> /checkout/tests/run-make/cross-lang-lto-clang/rmake.rs:13:29
   |
13 | const ALWAYS_INLINED_PATTERN = "bl.*<rust_always_inlined>";
   |
help: provide a type for the item
   |
   |
13 | const ALWAYS_INLINED_PATTERN: <type> = "bl.*<rust_always_inlined>";

error: missing type for `const` item
##[error]  --> /checkout/tests/run-make/cross-lang-lto-clang/rmake.rs:15:29
   |
   |
15 | const ALWAYS_INLINED_PATTERN = "call.*rust_always_inlined";
   |
help: provide a type for the item
   |
   |
15 | const ALWAYS_INLINED_PATTERN: <type> = "call.*rust_always_inlined";

error: missing type for `const` item
##[error]  --> /checkout/tests/run-make/cross-lang-lto-clang/rmake.rs:18:28
   |
   |
18 | const NEVER_INLINED_PATTERN = "bl.*<rust_never_inlined>";
   |
help: provide a type for the item
   |
   |
18 | const NEVER_INLINED_PATTERN: <type> = "bl.*<rust_never_inlined>";

error: missing type for `const` item
##[error]  --> /checkout/tests/run-make/cross-lang-lto-clang/rmake.rs:20:28
   |
   |
20 | const NEVER_INLINED_PATTERN = "call.*rust_never_inlined";
   |
help: provide a type for the item
   |
   |
20 | const NEVER_INLINED_PATTERN: <type> = "call.*rust_never_inlined";


fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/no-builtins-lto/rmake.rs" "/checkout/tests/run-make/no-builtins-lto/no_builtins.rs" "/checkout/tests/rustdoc-json/structs/tuple.rs" "/checkout/tests/rustdoc-json/structs/with_generics.rs" "/checkout/tests/rustdoc-json/return_private.rs" "/checkout/tests/run-make/box-struct-no-segfault/main.rs" "/checkout/tests/run-make/box-struct-no-segfault/rmake.rs" "/checkout/tests/run-make/box-struct-no-segfault/foo.rs" "/checkout/tests/rustdoc-json/lifetime/outlives_in_where.rs" "/checkout/tests/rustdoc-json/lifetime/outlives.rs" "/checkout/tests/rustdoc-json/lifetime/outlives_in_param.rs" "/checkout/tests/rustdoc-json/lifetime/longest.rs" "/checkout/tests/rustdoc-json/assoc_type.rs" "/checkout/tests/rustdoc-json/stripped_modules.rs" "/checkout/tests/rustdoc-json/doc_hidden_failure.rs" "/checkout/tests/rustdoc-json/keyword.rs" "/checkout/tests/rustdoc-json/generic-associated-types/gats.rs" "/checkout/tests/rustdoc-json/glob_import.rs" "/checkout/tests/rustdoc-json/impl-trait-in-assoc-type.rs" "/checkout/tests/rustdoc-json/output_generics.rs" "/checkout/tests/rustdoc-json/impl-trait-precise-capturing.rs" "/checkout/tests/rustdoc-json/non_lifetime_binders.rs" "/checkout/tests/run-make/checksum-freshness/rmake.rs" "/checkout/tests/run-make/checksum-freshness/foo.rs" "/checkout/tests/run-make/checksum-freshness/lib.rs" "/checkout/tests/run-make/extern-fn-struct-passing-abi/rmake.rs" "/checkout/tests/run-make/extern-fn-struct-passing-abi/test.rs" "/checkout/tests/run-make/forced-unwind-terminate-pof/rmake.rs" "/checkout/tests/run-make/forced-unwind-terminate-pof/foo.rs" "/checkout/tests/run-make/alloc-no-sync/rmake.rs" "/checkout/tests/run-make/export-executable-symbols/main.rs" "/checkout/tests/run-make/export-executable-symbols/rmake.rs" "/checkout/tests/rustdoc-json/impls/pub_for_hidden_private.rs" "/checkout/tests/rustdoc-json/impls/blanket_with_local.rs" "/checkout/tests/rustdoc-json/impls/impl_item_visibility_show_private.rs" "/checkout/tests/rustdoc-json/impls/auxiliary/foreign_trait.rs" "/checkout/tests/rustdoc-json/impls/auxiliary/foreign_struct.rs" "/checkout/tests/rustdoc-json/impls/issue-112852-dangling-trait-impl-id-3.rs" "/checkout/tests/rustdoc-json/impls/auto.rs" "/checkout/tests/rustdoc-json/impls/import_from_private.rs" "/checkout/tests/rustdoc-json/impls/local_for_primitive.rs" "/checkout/tests/rustdoc-json/impls/foreign_for_local.rs" "/checkout/tests/rustdoc-json/impls/local_for_foreign.rs" "/checkout/tests/rustdoc-json/impls/local_for_local_primitive.rs" "/checkout/tests/rustdoc-json/impls/impl_item_visibility.rs" "/checkout/tests/rustdoc-json/impls/impl_item_visibility_show_hidden.rs" "/checkout/tests/rustdoc-json/impls/issue-112852-dangling-trait-impl-id.rs" "/checkout/tests/rustdoc-json/impls/issue-112852-dangling-trait-impl-id-2.rs" "/checkout/tests/run-make/cross-lang-lto-clang/main.rs" "/checkout/tests/rustdoc-json/impls/local_for_local.rs" "/checkout/tests/run-make/cross-lang-lto-clang/rustlib.rs" "/checkout/tests/run-make/cross-lang-lto-clang/rmake.rs" "/checkout/tests/run-make/doctests-runtool/runtool.rs" "/checkout/tests/run-make/doctests-runtool/rmake.rs" "/checkout/tests/run-make/doctests-runtool/t.rs" "/checkout/tests/rustdoc-json/type/inherent_associated_type_bound.rs" "/checkout/tests/rustdoc-json/type/extern.rs" "/checkout/tests/rustdoc-json/type/generic_default.rs" "/checkout/tests/rustdoc-json/type/hrtb.rs" "/checkout/tests/rustdoc-json/type/inherent_associated_type.rs" "/checkout/tests/rustdoc-json/type/fn_lifetime.rs" "/checkout/tests/rustdoc-json/type/dyn.rs" "/checkout/tests/rustdoc-json/type/inherent_associated_type_projections.rs" "/checkout/tests/run-make/no-builtins-lto/main.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Oct 22 12:54:20 UTC 2024
  network time: Tue, 22 Oct 2024 12:54:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/ci/docker/host-aarch64/aarch64-gnu-debug/Dockerfile
+++ b/src/ci/docker/host-aarch64/aarch64-gnu-debug/Dockerfile
@@ -1,0 +1,58 @@
+FROM ubuntu:22.04
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get install -y --no-install-recommends \
+  g++ \
+  make \
+  ninja-build \
+  file \
+  curl \
+  ca-certificates \
+  python3 \
+  python3-dev \
+  libxml2-dev \
+  libncurses-dev \
+  libedit-dev \
+  swig \
+  doxygen \
+  git \
+  cmake \
+  sudo \
+  gdb \
+  libssl-dev \
+  pkg-config \
+  xz-utils \
+  lld \
+  clang \
+  && rm -rf /var/lib/apt/lists/*
+
+COPY scripts/sccache.sh /scripts/
+RUN sh /scripts/sccache.sh
+
+ENV RUSTBUILD_FORCE_CLANG_BASED_TESTS 1
+ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
+
+# llvm.use-linker conflicts with downloading CI LLVM
+ENV NO_DOWNLOAD_CI_LLVM 1
+
+ENV RUST_CONFIGURE_ARGS \
+      --build=aarch64-unknown-linux-gnu \
+      --enable-debug \
+      --enable-lld \
+      --set llvm.use-linker=lld \
+      --set target.aarch64-unknown-linux-gnu.linker=clang \
+      --set target.aarch64-unknown-linux-gnu.cc=clang \
+      --set target.aarch64-unknown-linux-gnu.cxx=clang++
+
+# This job appears to be checking two separate things:
+# - That we can build the compiler with `--enable-debug`
+#   (without necessarily testing the result).
+# - That the tests with `//@ needs-force-clang-based-tests` pass, since they
+#   don't run by default unless RUSTBUILD_FORCE_CLANG_BASED_TESTS is set.
+#   - FIXME(https://github.com/rust-lang/rust/pull/126155#issuecomment-2156314273):
+#     Currently we only run the subset of tests with "clang" in their name.
+#   - See also FIXME(#132034)
+
+ENV SCRIPT \
+  python3 ../x.py --stage 2 build && \
+  python3 ../x.py --stage 2 test tests/run-make --test-args clang

--- a/src/ci/github-actions/jobs.yml
+++ b/src/ci/github-actions/jobs.yml
@@ -123,6 +123,9 @@ auto:
   - image: aarch64-gnu
     <<: *job-aarch64-linux
 
+  - image: aarch64-gnu-debug
+    <<: *job-aarch64-linux
+
   - image: arm-android
     <<: *job-linux-4c
 

--- a/tests/run-make/cross-lang-lto-clang/rmake.rs
+++ b/tests/run-make/cross-lang-lto-clang/rmake.rs
@@ -9,6 +9,24 @@
 
 use run_make_support::{clang, env_var, llvm_ar, llvm_objdump, rustc, static_lib_name};
 
+#[cfg(any(target_arch = "aarch64", target_arch = "arm"))]
+static RUST_ALWAYS_INLINED_PATTERN: &'static str = "bl.*<rust_always_inlined>";
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+static RUST_ALWAYS_INLINED_PATTERN: &'static str = "call.*rust_always_inlined";
+#[cfg(any(target_arch = "aarch64", target_arch = "arm"))]
+static C_ALWAYS_INLINED_PATTERN: &'static str = "bl.*<c_always_inlined>";
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+static C_ALWAYS_INLINED_PATTERN: &'static str = "call.*c_always_inlined";
+
+#[cfg(any(target_arch = "aarch64", target_arch = "arm"))]
+static RUST_NEVER_INLINED_PATTERN: &'static str = "bl.*<rust_never_inlined>";
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+static RUST_NEVER_INLINED_PATTERN: &'static str = "call.*rust_never_inlined";
+#[cfg(any(target_arch = "aarch64", target_arch = "arm"))]
+static C_NEVER_INLINED_PATTERN: &'static str = "bl.*<c_never_inlined>";
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+static C_NEVER_INLINED_PATTERN: &'static str = "call.*c_never_inlined";
+
 fn main() {
     rustc()
         .linker_plugin_lto("on")
@@ -31,14 +49,14 @@ fn main() {
         .disassemble()
         .input("cmain")
         .run()
-        .assert_stdout_not_contains_regex("call.*rust_always_inlined");
+        .assert_stdout_not_contains_regex(RUST_ALWAYS_INLINED_PATTERN);
     // As a sanity check, make sure we do find a call instruction to a
     // non-inlined function
     llvm_objdump()
         .disassemble()
         .input("cmain")
         .run()
-        .assert_stdout_contains_regex("call.*rust_never_inlined");
+        .assert_stdout_contains_regex(RUST_NEVER_INLINED_PATTERN);
     clang().input("clib.c").lto("thin").arg("-c").out_exe("clib.o").arg("-O2").run();
     llvm_ar().obj_to_ar().output_input(static_lib_name("xyz"), "clib.o").run();
     rustc()
@@ -53,10 +71,10 @@ fn main() {
         .disassemble()
         .input("rsmain")
         .run()
-        .assert_stdout_not_contains_regex("call.*c_always_inlined");
+        .assert_stdout_not_contains_regex(C_ALWAYS_INLINED_PATTERN);
     llvm_objdump()
         .disassemble()
         .input("rsmain")
         .run()
-        .assert_stdout_contains_regex("call.*c_never_inlined");
+        .assert_stdout_contains_regex(C_NEVER_INLINED_PATTERN);
 }

--- a/tests/run-make/pointer-auth-link-with-c-lto-clang/rmake.rs
+++ b/tests/run-make/pointer-auth-link-with-c-lto-clang/rmake.rs
@@ -1,0 +1,36 @@
+// `-Z branch protection` is an unstable compiler feature which adds pointer-authentication
+// code (PAC), a useful hashing measure for verifying that pointers have not been modified.
+// This test checks that compilation and execution is successful when this feature is activated,
+// with some of its possible extra arguments (bti, pac-ret, leaf) when doing LTO.
+// See https://github.com/rust-lang/rust/pull/88354
+
+//@ needs-force-clang-based-tests
+//@ only-aarch64
+// Reason: branch protection is not supported on other architectures
+//@ ignore-cross-compile
+// Reason: the compiled binary is executed
+
+use run_make_support::{clang, env_var, llvm_ar, run, rustc, static_lib_name};
+
+fn main() {
+    clang()
+        .arg("-v")
+        .lto("thin")
+        .arg("-mbranch-protection=bti+pac-ret+leaf")
+        .arg("-O2")
+        .arg("-c")
+        .out_exe("test.o")
+        .input("test.c")
+        .run();
+    llvm_ar().obj_to_ar().output_input(static_lib_name("test"), "test.o").run();
+    rustc()
+        .linker_plugin_lto("on")
+        .opt_level("2")
+        .linker(&env_var("CLANG"))
+        .link_arg("-fuse-ld=lld")
+        .arg("-Zbranch-protection=bti,pac-ret,leaf")
+        .input("test.rs")
+        .output("test.bin")
+        .run();
+    run("test.bin");
+}

--- a/tests/run-make/pointer-auth-link-with-c-lto-clang/test.c
+++ b/tests/run-make/pointer-auth-link-with-c-lto-clang/test.c
@@ -1,0 +1,1 @@
+int foo() { return 0; }

--- a/tests/run-make/pointer-auth-link-with-c-lto-clang/test.rs
+++ b/tests/run-make/pointer-auth-link-with-c-lto-clang/test.rs
@@ -1,0 +1,10 @@
+#[link(name = "test")]
+extern "C" {
+    fn foo() -> i32;
+}
+
+fn main() {
+    unsafe {
+        foo();
+    }
+}

